### PR TITLE
Add missing complex support for torch.norm and torch.linalg.norm

### LINKED ISSUE #47833: Support `torch.norm` for complex tensors on both CPU and CUDA
## Motivation

`torch.norm` is a commonly used operation. TorchAudio and other libraries/users can benefit from the performance gains if `torch.norm` is supported. https://github.com/pytorch/audio/pull/1009#discussion_r518965072

## Pitch

We should 
- support `torch.norm` for complex tensors whenever the operation makes sense and error out otherwise (p=inf, -inf).
- update the backward formula using the complex autograd guide here: https://github.com/pytorch/pytorch/issues/33152

### PR DESCRIPTION
**Pull request by kurtamohler**
**BC-breaking note:** 

Previously, when given a complex input, `torch.linalg.norm` and `torch.norm` would return a complex output. `torch.linalg.cond` would sometimes return a complex output and sometimes return a real output when given a complex input, depending on its `p` argument. This PR changes this behavior to match `numpy.linalg.norm` and `numpy.linalg.cond`, so that a complex input will result in the downgraded real number type, consistent with NumPy.

**PR Summary:**

The following cases were previously unsupported for complex inputs, and this commit adds support:

- Frobenius norm
- Norm order 2 (vector and matrix)
- CUDA vector norm

Part of #47833

### CODE REVIEW

File: test/test_linalg.py [hunk 1]
```diff
@@ -1062,57 +1062,36 @@ def run_error_test_case(input, ord, dim, keepdim, error_type, error_regex):
...
+                torch.linalg.norm(x, ord, keepdim=keepdim, out=res_out)
+                self.assertEqual(res_out.shape, expected.shape, msg=msg)
+                self.assertEqual(res_out.cpu(), expected, msg=msg)
+
             # matrix norm
             x = torch.randn(25, 25, device=device, dtype=dtype)
             xn = x.cpu().numpy()
-            for ord in supported_matrix_ords:
+            for ord in matrix_ords:
                 # TODO: Need to fix abort when nuclear norm is given cdouble input:
```
Comment (kurtamohler) [inline test/test_linalg.py:1095]: Note to self, this still needs to be addressed

**Comment by kurtamohler** [comment]
Rocm test failure seems to be real

File: aten/src/ATen/native/SharedReduceOps.h [hunk 2]
```diff
@@ -157,15 +150,32 @@ struct MeanOps {
...
 };
 
+#if defined(__CUDACC__) || defined(__HIPCC__)
+template <typename T>
+inline C10_DEVICE T min_impl(T a, T b) {
+  return ::min(a, b);
+}
+
+template <>
+inline C10_DEVICE c10::complex<float> min_impl(c10::complex<float> a, c10::complex<float> b) {
```
Comment (kurtamohler) [inline aten/src/ATen/native/SharedReduceOps.h:160]: An implicit assumption in these complex min/max/pow/sqrt functions is that the imaginary part of the complex numbers is 0, since we are always calling `abs` on them first. This assumption is required for numpy compatibility, and it also circumvents the fact that you can't actually perform a true min/max (if inputs' norms are equal but the inputs are not) or a pow (for instance, square root will usually have two correct answers) with complex numbers.

Perhaps it would be better to just take out the real part and operate upon it. If we do that, it might be good to have an assert to check that the imaginary part is 0. It would be a device side assert, which isn't great, but it should never actually get triggered by anything a user can do. The assert would just ensure that developers will only call this on complex numbers that have been normed.

But I think there's an even better solution. Since these AbsMinOps/AbsMaxOps/etc accumulators need to accumulate the absolute values of complex numbers, it would be much better to accumulate them as the downgraded type. This would increase performance and we would avoid needing to use min/max/pow/sqrt functions that accept complex numbers. I'll go ahead and implement this solution, unless anyone has a better suggestion.

cc @mruberry

**Comment by kurtamohler** [comment]
@mruberry, in numpy, if we give norm a complex number, it returns the downgraded real type:
```
>>> numpy.linalg.norm([1+1j]).dtype
dtype('float64')
```
But in pytorch, we match the input type:
```
>>> torch.linalg.norm(torch.tensor([1+1j])).dtype
torch.complex64
```

Is it alright if we BC-break to fix this?

File: aten/src/ATen/native/SharedReduceOps.h (hunk 2 above)
Comment (kurtamohler) [reply aten/src/ATen/native/SharedReduceOps.h]: I've updated the norm kernels to use the downgraded scalar type for the accumulators.

File: aten/src/ATen/native/ReduceOps.cpp [hunk 3]
```diff
@@ -660,8 +660,7 @@ Tensor& logsumexp_out(Tensor& result, const Tensor& self, DimnameList dims, bool
 
 static Tensor& norm_out(Tensor &result, const Tensor &self, optional<Scalar> opt_p,
                                IntArrayRef dim, bool keepdim, optional<ScalarType> opt_dtype) {
-  auto p = opt_p.value_or(2.0);
-  TORCH_CHECK(!(p.toDouble() == 2 && self.is_complex()), "norm with p=2 not supported for complex tensors");
+  auto p = opt_p.value_or(2.0).to<double>();
```
Comment (mruberry) [inline aten/src/ATen/native/ReduceOps.cpp:663]: What happens if opt_p is complex?

I imagine a lot of our code is casting to double naively.

File: aten/src/ATen/native/SharedReduceOps.h [hunk 4]
```diff
@@ -203,18 +203,18 @@ struct AbsMaxOps {
...
 
-template <typename acc_t>
+template <typename scalar_t, typename acc_t=scalar_t>
 struct NormOps {
   acc_t norm_;
 
-  inline C10_DEVICE acc_t reduce(acc_t acc, acc_t data, int64_t /*idx*/) const {
-    return acc + compat_pow(std::abs(data), norm_);
+  inline C10_DEVICE acc_t reduce(acc_t acc, scalar_t data, int64_t /*idx*/) const {
+    return acc + compat_pow(acc_t(std::abs(data)), norm_);
```
Comment (mruberry) [inline aten/src/ATen/native/SharedReduceOps.h:217]: What does the `acc_t` do here? Is it equivalent to `static_cast<acc_t>(...)`?

File: aten/src/ATen/native/cpu/ReduceOpsKernel.cpp [hunk 5]
```diff
@@ -174,61 +174,72 @@ static void norm_kernel_tensor_iterator_impl(
   if (p.isIntegral(false)) {
     val = p.to<int64_t>();
   } else if (p.isFloatingPoint()) {
-    val = p.to<float>();
+    val = p.to<double>();
```
Comment (mruberry) [inline aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:177]: Why the change from float to double here?

File: aten/src/ATen/native/SharedReduceOps.h [hunk 6]
```diff
@@ -157,10 +157,10 @@ struct MeanOps {
   }
 };
 
-template <typename acc_t>
+template <typename scalar_t, typename acc_t=scalar_t>
```
Comment (mruberry) [inline aten/src/ATen/native/SharedReduceOps.h:160]: Add a comment here explaining the template and its intended use.

The use of `acc_t` seems accurate here but is different from the typical notion of an accumulate type:

https://github.com/pytorch/pytorch/blob/324c18fcad579b1afa63ae45528bf598ba8ec4ca/aten/src/ATen/AccumulateType.h#L21

This PR should add comments as needed to make this distinction clear to avoid confusion.

File: aten/src/ATen/native/SharedReduceOps.h [hunk 7]
```diff
@@ -284,16 +284,17 @@ struct NormOneOps {
...
 #endif
 };
 
-template <typename acc_t>
+template <typename scalar_t, typename acc_t=scalar_t>
 struct NormTwoOps {
-  inline C10_DEVICE acc_t reduce(acc_t acc, acc_t data, int64_t /*idx*/) const {
-    return acc + data * data;
+  inline C10_DEVICE acc_t reduce(acc_t acc, scalar_t data, int64_t /*idx*/) const {
+    acc_t abs_data = std::abs(data);
```
Comment (mruberry) [inline aten/src/ATen/native/SharedReduceOps.h:296]: What's the difference between the previous expression and the new one with `std::abs`?

File: aten/src/ATen/native/cpu/ReduceOpsKernel.cpp [hunk 8]
```diff
@@ -174,61 +174,72 @@ static void norm_kernel_tensor_iterator_impl(
...
   } else {
     AT_ERROR("norm_kernel_tensor_iterator_impl expects norm to be integer or float");
   }
 
-
   if (val == 0) {
     AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND2(kHalf, kBFloat16, iter.dtype(), "norm_cpu", [&] {
+      using acc_t = typename scalar_value_type<scalar_t>::type;
```
Comment (mruberry) [inline aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:184]: Another `acc_t` comment (see above) should probably be added here, too.

File: aten/src/ATen/native/cpu/ReduceOpsKernel.cpp [hunk 9]
```diff
@@ -174,61 +174,72 @@ static void norm_kernel_tensor_iterator_impl(
...
         iter,
-        NormOps<scalar_t> { scalar_t(val) },
-        scalar_t(0)
+        NormOps<scalar_t, acc_t> { acc_t(val) },
+        acc_t(0)
       );
     });
   }
+
+  // For complex outputs, the above kernels do not touch the imaginary values,
```
Comment (mruberry) [inline aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:238]: Would you elaborate on this? How do the above kernels not edit the imaginary values of output?

I wonder if it's faster to zero output before it's modified instead of zeroing just the imaginary part after its been modified.

File: aten/src/ATen/native/cuda/ReduceNormKernel.cu [hunk 10]
```diff
@@ -7,29 +7,29 @@
 
 namespace at { namespace native {
 
-template <typename scalar_t, typename acc_t=scalar_t, typename out_t=scalar_t>
+template <typename scalar_t, typename acc_t=typename scalar_value_type<scalar_t>::type, typename out_t=scalar_t>
 void norm_kernel_cuda_impl(TensorIterator& iter, Scalar val) {
   float p;
```
Comment (mruberry) [inline aten/src/ATen/native/cuda/ReduceNormKernel.cu:12]: Should this be double p?

**Comment by mruberry** [comment]
Hey @kurtamohler!

This PR looks great. It has a surgical precision. I made a few comments that are mostly me asking questions to better understand what's going on.

I think this PR should update the documentation, too, to more accurately describe torch.norm's and torch.linalg.norm's complex support.

Also, how are we testing complex autograd for torch.linalg.norm? Should something like this be updated?

https://github.com/pytorch/pytorch/blob/bb5d4984b912f3f9f775b00b31de198fd3d01a7f/test/test_linalg.py#L937

File: aten/src/ATen/native/ReduceOps.cpp (hunk 3 above)
Comment (kurtamohler) [reply aten/src/ATen/native/ReduceOps.cpp:663]: Looks like attempting to cast complex to double throws an error:
```
>>> torch.linalg.norm(torch.randn(10), ord=1j)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: value cannot be converted to type double without overflow: (0,1)
```

File: aten/src/ATen/native/SharedReduceOps.h (hunk 4 above)
Comment (kurtamohler) [reply aten/src/ATen/native/SharedReduceOps.h:217]: Yeah this is a C-style cast. But it's probably better to use `static_cast` (https://anteru.net/blog/2007/c-background-static-reinterpret-and-c-style-casts/), so I'll go ahead and change all of these.

File: aten/src/ATen/native/cpu/ReduceOpsKernel.cpp (hunk 5 above)
Comment (kurtamohler) [reply aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:177]: To me, it seems better to cast to double to allow more precision for `p`.

File: aten/src/ATen/native/SharedReduceOps.h (hunk 7 above)
Comment (kurtamohler) [reply aten/src/ATen/native/SharedReduceOps.h:296]: The `std::abs()` call is needed when `scalar_t` is complex. For real numbers, it's not needed. I could add an overload for complex numbers so that we avoid calling `std::abs()` for real numbers.

File: aten/src/ATen/native/cpu/ReduceOpsKernel.cpp (hunk 9 above)
Comment (kurtamohler) [reply aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:238]: Sure, I can elaborate. I'll try zeroing the whole thing first and see how it affects performance.

File: aten/src/ATen/native/cuda/ReduceNormKernel.cu (hunk 10 above)
Comment (kurtamohler) [reply aten/src/ATen/native/cuda/ReduceNormKernel.cu:12]: Yeah, I think that is a good idea

**Comment by kurtamohler** [comment]
Thanks for the review @mruberry! Yes, more accurate documentation is a good idea, and I will look into autograd testing

File: aten/src/ATen/native/SharedReduceOps.h (hunk 7 above)
Comment (mruberry) [reply aten/src/ATen/native/SharedReduceOps.h:296]: Adding an overload sounds good as long as it doesn't add too much code complexity.

File: aten/src/ATen/native/cpu/ReduceOpsKernel.cpp (hunk 9 above)
Comment (kurtamohler) [reply aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:238]: Actually this special case won't be necessary any more after I change the output types to match numpy.

Comment (kurtamohler) [reply aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:238]: Nevermind, I misspoke. This is special case is still important if `out`'s dtype or the `dtype` argument is specified as complex.

File: aten/src/ATen/native/SharedReduceOps.h (hunk 7 above)
Comment (kurtamohler) [reply aten/src/ATen/native/SharedReduceOps.h]: I added `abs_if_complex()`. Not too sure if this is the most efficient way to implement it. Let me know if you think this adds too much complexity.

File: test/test_linalg.py [hunk 11]
```diff
@@ -1083,72 +1121,95 @@ def gen_error_message(input_size, ord, keepdim, dim=None):
...
+            dtype_real = torch.float
+        elif dtype == torch.cdouble:
+            dtype_real = torch.double
+        else:
+            raise RuntimeError(f'dtype not supported in this test: {dtype}')
+
+        vector_ords = [None, 0, 1, 2, 3, inf, -1, -2, -3, -inf]
+        matrix_ords = [None, 'fro', 1, inf, -1, -inf]
+
+        # TODO: Fix autograd for matrix orders 'nuc', 2, and -2 by adding complex
```
Comment (kurtamohler) [inline test/test_linalg.py:1178]: @mruberry, I figure adding complex support to `svd`'s backward method should probably be done in a separate PR.

File: aten/src/ATen/native/cpu/ReduceOpsKernel.cpp (hunk 9 above)
Comment (kurtamohler) [reply aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:241]: Ok, I looked into whether zeroing out the entire tensor before the norm calculation would be faster than zeroing out only the imaginary part after the norm calculation. I used this script: https://github.com/kurtamohler/pytorch-perf-test-scripts/blob/master/linalg_norm/run_single_core_norm_output_complex_perf.sh

There doesn't seem to be a significant difference between the two options.

**Comment by kurtamohler** [comment]
@mruberry, I've addressed all your comments and made the changes you requested. I think it's ready for a re-review

File: aten/src/ATen/native/cpu/ReduceOpsKernel.cpp (hunk 9 above)
Comment (mruberry) [reply aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:241]: Cool. Thanks for checking, @kurtamohler!

File: test/test_linalg.py (hunk 11 above)
Comment (mruberry) [reply test/test_linalg.py:1178]: That sounds great.

File: test/test_linalg.py [hunk 12]
```diff
@@ -825,23 +825,61 @@ def test_kron_errors_and_warnings(self, device, dtype):
     # as expected, according to the function's documentation
     @skipCUDAIfNoMagma
     def test_norm_dtype(self, device):
-        def run_test_case(input_size, ord, keepdim, from_dtype, to_dtype, compare_dtype):
+        def run_test_case(input_size, ord, keepdim, from_dtype, to_dtype):
+            # Determine the best dtype to use for comparisons between tensors
+            # of two different types
+            def get_compare_dtype(type0, type1):
+                types_32bit_based = [torch.float, torch.cfloat]
```
Comment (mruberry) [inline test/test_linalg.py:832]: can get_compare_dtype just be torch.promote_types()? 

https://pytorch.org/docs/master/generated/torch.promote_types.html?highlight=promote#torch.promote_types

File: test/test_linalg.py [hunk 13]
```diff
@@ -825,23 +825,61 @@ def test_kron_errors_and_warnings(self, device, dtype):
...
+                is_complex = type0.is_complex or type1.is_complex
+
+                if type0 in types_32bit_based or type1 in types_32bit_based:
+                    return torch.cfloat if is_complex else torch.float
+                else:
+                    return torch.cdouble if is_complex else torch.double
+
+            compare_dtype = get_compare_dtype(from_dtype, to_dtype)
+
+            def get_value_type(dtype):
```
Comment (mruberry) [inline test/test_linalg.py:842]: @anjali411 We should think about a Python exposure for acquiring the value type. Maybe we just want to add this as a torch.testing helper function in the future.

File: torch/functional.py [hunk 14]
```diff
@@ -1262,7 +1262,12 @@ def norm(input, p="fro", dim=None, keepdim=False, out=None, dtype=None):  # noqa
         more consistent with NumPy's numpy.linalg.norm.
 
     Args:
-        input (Tensor): the input tensor
+        input (Tensor): The input tensor. The data type must be either floating
```
Comment (mruberry) [inline torch/functional.py:1265]: "The data type must be either floating point or complex" -> "Its data type must be either a floating point or complex type."

File: torch/functional.py [hunk 15]
```diff
@@ -1262,7 +1262,12 @@ def norm(input, p="fro", dim=None, keepdim=False, out=None, dtype=None):  # noqa
...
+            point or complex. For complex inputs, the norm is calculated on the
+            absolute values of each element. If the input is complex and neither
+            :attr:`dtype` nor :attr:`out` is specified, the result's data type will
+            be the corresponding downgraded real number type.
```
Comment (mruberry) [inline torch/functional.py:1269]: "corresponding downgraded real number type" -> "corresponding floating point type (e.g. float if :attr:\`input\` is complexfloat)."

File: torch/linalg/__init__.py [hunk 16]
```diff
@@ -286,7 +286,10 @@
 Args:
     input (Tensor): The input tensor. If dim is None, x must be 1-D or 2-D, unless :attr:`ord`
         is None. If both :attr:`dim` and :attr:`ord` are None, the 2-norm of the input flattened to 1-D
-        will be returned.
+        will be returned. The data type must be either floating point or complex. For complex inputs,
```
Comment (mruberry) [inline torch/linalg/__init__.py:289]: Two comments above apply to this snippet, too.

**Review by mruberry — APPROVED** [review]
Awesome! Thanks @kurtamohler.

Two small cleanup suggestions in the docs and one question about using torch.promote_types() to simplify a test. Just let me know when this is ready to merge.

File: test/test_linalg.py (hunk 12 above)
Comment (kurtamohler) [reply test/test_linalg.py:832]: `torch.promote_types` doesn't seem to do what I want. It chooses the higher precision type and prefers complex. But I want to choose the lower precision type and prefer complex.

Comment (mruberry) [reply test/test_linalg.py:832]: Aha, thanks for taking a look. My mistake.

**Comment by kurtamohler** [comment]
@mruberry, I think this is ready to merge

**Comment by kurtamohler** [comment]
Looks like the return dtype change (return real when given complex) is breaking one of the tests for `torch.linalg.cond` because it depends on `torch.linalg.norm`. In numpy, `cond` has the same return type behavior as `norm`:

```
>>> a = numpy.random.rand(10, 10) + 1j * numpy.random.rand(10, 10)
>>> a.dtype
dtype('complex128')
>>> numpy.linalg.cond(a).dtype
dtype('float64')
```

So I can update the failing test and add `cond` to the BC breaking note above.

**Comment by kurtamohler** [comment]
I'm not sure what's causing the `pytorch-linux-bionic-rocm3.9-py3.6` failure. I'll look into it

**Comment by mruberry** [comment]
> I'm not sure what's causing the `pytorch-linux-bionic-rocm3.9-py3.6` failure. I'll look into it

It doesn't look related. We can probably ignore it (unless the same failure happens again).

**Comment by kurtamohler** [comment]
In that case, I think this is ready to merge @mruberry , unless another test fails after I rebased

**Comment by kurtamohler** [comment]
All the failures are due to an upstream mypy issue

File: aten/src/ATen/native/ReduceOps.cpp [hunk 17]
```diff
@@ -660,22 +660,23 @@ Tensor& logsumexp_out(Tensor& result, const Tensor& self, DimnameList dims, bool
...
               "norm only supports CPU AND CUDA device type, got: ", self.device().type());
   TORCH_CHECK(self.layout() == Layout::Strided,
               "norm only supports strided layout, got: ", self.layout());
 
-  ScalarType scalarType = opt_dtype.has_value() ? opt_dtype.value() : self.scalar_type();
+  ScalarType in_dtype = opt_dtype.has_value() ? opt_dtype.value() : self.scalar_type();
   TORCH_CHECK(
-      at::isFloatingType(scalarType) || at::isComplexType(scalarType),
+      at::isFloatingType(in_dtype) || at::isComplexType(in_dtype),
       "Can only calculate the mean of floating types. Got ",
```
Comment (mruberry) [inline aten/src/ATen/native/ReduceOps.cpp:672]: This error message seems odd. Should it be:

"Can only calculate the norm of floating point and complex dtypes..." ?

File: torch/functional.py [hunk 18]
```diff
@@ -1293,7 +1293,13 @@ def norm(input, p="fro", dim=None, keepdim=False, out=None, dtype=None):  # noqa
         more consistent with NumPy's numpy.linalg.norm.
 
     Args:
-        input (Tensor): the input tensor
+        input (Tensor): The input tensor. Its data type must be either a floating
+            point or complex type. For complex inputs, the norm is calculated on the
+            absolute values of each element. If the input is complex and neither
```
Comment (mruberry) [inline torch/functional.py:1298]: "calculated on the absolute values" -> "calculated using the absolute value"

File: torch/linalg/__init__.py [hunk 19]
```diff
@@ -286,7 +286,10 @@
 Args:
     input (Tensor): The input tensor. If dim is None, x must be 1-D or 2-D, unless :attr:`ord`
         is None. If both :attr:`dim` and :attr:`ord` are None, the 2-norm of the input flattened to 1-D
-        will be returned.
+        will be returned. Its data type must be either a floating point or complex type. For complex
+        inputs, the norm is calculated on of the absolute values of each element. If the input is
+        complex and neither :attr:`dtype` nor :attr:`out` is specified, the result's data type will be
+        the corresponding downgraded real number type.
```
Comment (mruberry) [inline torch/linalg/__init__.py:292]: This sentence needs to be updated to be consistent with the torch/functional.py documentation.

Comment (mruberry) [reply torch/linalg/__init__.py:292]: Does the torch.linalg.cond documentation need a similar update, too?

**Comment by mruberry** [comment]
Hey @kurtamohler! Just a few comments/questions. Overall things look great.

File: torch/linalg/__init__.py (hunk 19 above)
Comment (kurtamohler) [reply torch/linalg/__init__.py]: Updated this and `torch.linalg.cond`

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 20]
```diff
@@ -1679,7 +1677,7 @@ static Tensor& _linalg_norm_vector_out(Tensor& result, const Tensor& self, optio
       // when the input contains extreme values (like nan or +/-inf) or if the input
       // size is degenerate (like size(0), size(0, N), etc)
       case_was_overridden = true;
-      self_ = self.abs();
+      self_ = self_.abs();
```
Comment (anjali411) [inline aten/src/ATen/native/LinearAlgebra.cpp:1680]: why don't we just declare self_ without assignment above?

Comment (kurtamohler) [reply aten/src/ATen/native/LinearAlgebra.cpp:1680]: The conditional dtype conversion `opt_dtype.has_value() ? self.to(opt_dtype.value()) : self` needs to be performed for all cases, which is the reason why `self_` is defined
